src/auth/jsonfile.rs

```diff
@@ -0,0 +1,69 @@
+use crate::auth::*;
+
+use futures::Future;
+
+use std::fs;
+use serde::Deserialize;
+use serde_json;
+
+use tokio::timer::Delay;
+use std::time::{Duration, Instant};
+
+#[derive(Deserialize, Clone, Debug)]
+struct Credentials {
+    username: String,
+    password: String,
+}
+
+/// [`Authenticator`] implementation that authenticates against a JSON file.
+///
+/// [`Authenticator`]: ../trait.Authenticator.html
+#[derive(Clone, Debug)]
+pub struct JsonfileAuthenticator {
+    credentials_list: Vec<Credentials>,
+}
+
+impl JsonfileAuthenticator {
+    /// Initialize a new [`JsonfileAuthenticator`] from file.
+    pub fn new(filename: String) -> Self {
+        let s = fs::read_to_string(filename).unwrap();
+        let credentials_list: Vec<Credentials> = serde_json::from_str(&s).unwrap();
```

Comment:
These unwraps are probably not good in the library. You could make the method return `Result<Self,Error>` and pass the errors to the caller for handling.

Reply:
Thanks for the suggestion. I kinda realized this was a bad idea to be done with it ;), but I was lacking inspiration on how to do this right last Friday night. Yes I'll propagate the error to the caller.